pkg/kubelet/status/status_manager.go

```diff
@@ -335,19 +337,26 @@ func (m *manager) TerminatePod(pod *v1.Pod) {
 		oldStatus = &cachedStatus.status
 	}
 	status := *oldStatus.DeepCopy()
-	for i := range status.ContainerStatuses {
-		if status.ContainerStatuses[i].State.Terminated != nil {
-			continue
-		}
-		status.ContainerStatuses[i].State = v1.ContainerState{
-			Terminated: &v1.ContainerStateTerminated{
-				Reason:   "ContainerStatusUnknown",
-				Message:  "The container could not be located when the pod was terminated",
-				ExitCode: 137,
-			},
+
+	// once a pod has initialized, any missing status is treated as a failure
```

Comment:
Changes here should preserve the invariant that at least one container is failed in the pod (Init or regular) or we mark all waiting containers as terminated with 137.  Reason being that hasPodInitialized is true if the last init container is successful or there are no init containers, OR if there is a failing init container (which for restart never pods is enough to mark as failure).

The e2e in #107829 should be sufficient to test this didn't regress, but I may resurrect a test of a similar behavior on init containers.

Reply:
> Changes here should preserve the invariant that at least one container is failed in the pod (Init or regular) or we mark all waiting containers as terminated with 137. 

The current code checks

```
if state := container.State; state.Terminated != nil && state.Terminated.ExitCode == 0 {
			return true
}
```

so if exitCode != 0, the `hasPodInitialized` will return false. This seems to be contradict the statement:

> hasPodInitialized is true if the last init container is successful or there are no init containers, **OR if there is a failing init container**

So in the current case, AFAIU, if there is failing init container, `hasPodInitialized` will return false, and we will not enter the body of the loop and set rest of container status to exit code 137? is that the expected behavior?